File: mycuration/src/main/java/com/phicdy/mycuration/presentation/view/fragment/CurationListFragment.java
package com.phicdy.mycuration.presentation.view.fragment;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import com.phicdy.mycuration.R;
import com.phicdy.mycuration.data.db.DatabaseAdapter;
import com.phicdy.mycuration.presentation.presenter.CurationListPresenter;
import com.phicdy.mycuration.data.rss.Curation;
import com.phicdy.mycuration.data.rss.UnreadCountManager;
import com.phicdy.mycuration.presentation.view.CurationListView;
import com.phicdy.mycuration.presentation.view.activity.AddCurationActivity;

import java.util.ArrayList;


public class CurationListFragment extends Fragment implements CurationListView {

    private CurationListPresenter presenter;
    private CurationListAdapter curationListAdapter;
    private OnCurationListFragmentListener mListener;
    private ListView curationListView;
    private TextView emptyView;

    private UnreadCountManager unreadManager;

    private static final int EDIT_CURATION_MENU_ID = 1;
    private static final int DELETE_CURATION_MENU_ID = 2;

    public CurationListFragment() {
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance();
        presenter = new CurationListPresenter(dbAdapter);
        presenter.setView(this);
        unreadManager = UnreadCountManager.getInstance();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.resume();
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);
        menu.clear();
        menu.add(0, EDIT_CURATION_MENU_ID, 0, R.string.edit_curation);
        menu.add(0, DELETE_CURATION_MENU_ID, 0, R.string.delete_curation);
    }

    public boolean onContextItemSelected(MenuItem item) {
        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();

        switch (item.getItemId()) {
            case EDIT_CURATION_MENU_ID:
                Curation editCuration = curationListAdapter.getItem(info.position);
                if (editCuration != null) {
                    presenter.onCurationEditClicked(editCuration.getId());
                }
                return true;
            case DELETE_CURATION_MENU_ID:
                Curation curation = curationListAdapter.getItem(info.position);
                presenter.onCurationDeleteClicked(curation);
                return true;
            default:
                return super.onContextItemSelected(item);
        }
    }

    private void setAllListener() {
        curationListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view,
                                    int position, long id) {
                Curation curation = curationAt(position);
                if (curation != null) {
                    mListener.onCurationListClicked(curation.getId());
                }
            }
        });
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_curation_list, container, false);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            mListener = (OnCurationListFragmentListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(context.toString()
                    + " must implement OnCurationListFragmentListener");
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        curationListView = (ListView) getActivity().findViewById(R.id.lv_curation);
        emptyView = (TextView)getActivity().findViewById(R.id.emptyView_curation);
        setAllListener();
        presenter.activityCreated();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void startEditCurationActivity(int editCurationId) {
        Intent intent = new Intent();
        intent.setClass(getActivity(), AddCurationActivity.class);
        intent.putExtra(AddCurationFragment.EDIT_CURATION_ID, editCurationId);
        startActivity(intent);
    }

    @Override
    public void setNoRssTextToEmptyView() {
        emptyView.setText(R.string.no_rss_message);
    }

    @Override
    public void setEmptyViewToList() {
        curationListView.setEmptyView(emptyView);
    }

    @Override
    public void registerContextMenu() {
        registerForContextMenu(curationListView);
    }

    @Override
    public void initListBy(ArrayList<Curation> curations) {
        curationListAdapter = new CurationListAdapter(curations, getActivity());
        curationListView.setAdapter(curationListAdapter);
        curationListAdapter.notifyDataSetChanged();
    }

    @Override
    public void delete(Curation curation) {
        curationListAdapter.remove(curation);
        curationListAdapter.notifyDataSetChanged();
    }

    @Override
    public int size() {
        if (curationListAdapter == null) return 0;
        return curationListAdapter.getCount();
    }

    @Override
    public Curation curationAt(int position) {
        if (curationListAdapter == null || position > curationListAdapter.getCount())
            return null;
        return curationListAdapter.getItem(position);
    }

    public interface OnCurationListFragmentListener {
        void onCurationListClicked(int curationId);
    }

    class CurationListAdapter extends ArrayAdapter<Curation> {
        CurationListAdapter(ArrayList<Curation> curations, Context context) {
            super(context, R.layout.curation_list, curations);
        }

        @NonNull
        @Override
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            ViewHolder holder;

            // Use contentView and setup ViewHolder
            View row = convertView;
            if (convertView == null) {
                LayoutInflater inflater = getActivity().getLayoutInflater();
                row = inflater.inflate(R.layout.curation_list, parent, false);
                holder = new ViewHolder();
                holder.curationName = (TextView) row.findViewById(R.id.tv_curation_title);
                holder.curationCount = (TextView) row.findViewById(R.id.tv_curation_count);
                row.setTag(holder);
            } else {
                holder = (ViewHolder) row.getTag();
            }

            Curation curation = this.getItem(position);
            if (curation != null) {
                holder.curationName.setText(curation.getName());
                holder.curationCount.setText(String.valueOf(unreadManager.getCurationCount(curation.getId())));
            }
            return row;
        }

        private class ViewHolder {
            TextView curationName;
            TextView curationCount;
        }
    }
}

File: mycuration/src/main/java/com/phicdy/mycuration/presentation/view/fragment/CurationListFragment.kt
package com.phicdy.mycuration.presentation.view.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.phicdy.mycuration.R
import com.phicdy.mycuration.data.rss.Curation
import com.phicdy.mycuration.presentation.presenter.CurationListPresenter
import com.phicdy.mycuration.presentation.view.CurationItem
import com.phicdy.mycuration.presentation.view.CurationListView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.koin.android.ext.android.inject
import org.koin.android.scope.ext.android.bindScope
import org.koin.android.scope.ext.android.getOrCreateScope
import org.koin.core.parameter.parametersOf
import java.util.ArrayList
import kotlin.coroutines.CoroutineContext


class CurationListFragment : Fragment(), CurationListView, CoroutineScope {

    companion object {
        private const val EDIT_CURATION_MENU_ID = 1
        private const val DELETE_CURATION_MENU_ID = 2
    }

    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    private val presenter: CurationListPresenter by inject { parametersOf(this) }
    private lateinit var curationListAdapter: CurationListAdapter
    private var mListener: OnCurationListFragmentListener? = null
    private lateinit var curationRecyclerView: RecyclerView
    private lateinit var emptyView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindScope(getOrCreateScope("curation_list"))
        job = Job()
    }

    override fun onResume() {
        super.onResume()
        presenter.resume()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_curation_list, container, false)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        try {
            mListener = context as OnCurationListFragmentListener
        } catch (e: ClassCastException) {
            throw ClassCastException(context.toString() + " must implement OnCurationListFragmentListener")
        }

    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        curationRecyclerView = activity?.findViewById(R.id.rv_curation) as RecyclerView
        emptyView = activity?.findViewById(R.id.emptyView_curation) as TextView
        launch {
            presenter.activityCreated()
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun onDestroy() {
        job.cancel()
        super.onDestroy()
    }

    override fun startEditCurationActivity(editCurationId: Int) {
        mListener?.startEditCurationActivity(editCurationId)
    }

    override fun setNoRssTextToEmptyView() {
        emptyView.setText(R.string.no_rss_message)
    }

    override fun registerContextMenu() {
        registerForContextMenu(curationRecyclerView)
    }

    override fun initListBy(curations: ArrayList<Curation>) {
        curationListAdapter = CurationListAdapter(curations)
        curationRecyclerView.adapter = curationListAdapter
        curationRecyclerView.layoutManager = LinearLayoutManager(activity)
        curationListAdapter.notifyDataSetChanged()
    }

    override fun showRecyclerView() {
        curationRecyclerView.visibility = View.VISIBLE
    }

    override fun hideRecyclerView() {
        curationRecyclerView.visibility = View.GONE
    }

    override fun showEmptyView() {
        emptyView.visibility = View.VISIBLE
    }

    override fun hideEmptyView() {
        emptyView.visibility = View.GONE
    }

    interface OnCurationListFragmentListener {
        fun onCurationListClicked(curationId: Int)
        fun startEditCurationActivity(editCurationId: Int)
    }

    inner class CurationListAdapter(private val curations: ArrayList<Curation>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
            val itemView = LayoutInflater.from(parent.context).inflate(R.layout.curation_list, parent, false)
            return ViewHolder(itemView)
        }

        override fun getItemCount(): Int {
            return curations.size
        }

        override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
            holder.itemView.setOnClickListener {
                val curation = curations[position]
                mListener?.onCurationListClicked(curation.id)
            }

            holder.itemView.setOnCreateContextMenuListener { menu, _, _ ->
                menu.add(0, EDIT_CURATION_MENU_ID, 0, R.string.edit_curation).setOnMenuItemClickListener {
                    val editCuration = curations[position]
                    presenter.onCurationEditClicked(editCuration.id)
                    true
                }
                menu.add(0, DELETE_CURATION_MENU_ID, 0, R.string.delete_curation).setOnMenuItemClickListener {
                    launch {
                        presenter.onCurationDeleteClicked(curations[position], curations.size)
                        curations.removeAt(position)
                        notifyItemRemoved(position)
                        delay(500) // Wait for delete animation
                        notifyDataSetChanged()
                    }
                    true
                }
            }
            val curation = curations[position]
            launch {
                presenter.getView(curation, holder as CurationItem)
            }

        }

        inner class ViewHolder(itemView: View): RecyclerView.ViewHolder(itemView), CurationItem {
            val curationName = itemView.findViewById(R.id.tv_curation_title) as TextView
            val curationCount = itemView.findViewById(R.id.tv_curation_count) as TextView

            override fun setName(name: String) {
                curationName.text = name
            }

            override fun setCount(count: String) {
                curationCount.text = count
            }
        }
    }
}
